trieste/models/model_interfaces.py

```diff
@@ -419,6 +420,60 @@ def get_observation_noise(self):
 
         return noise_variance
 
+    def optimize(self, dataset: Dataset) -> None:
+        """
+        Optimize the model with the specified `dataset`.
+
+        If any of the kernel's trainable parameters have priors, then we begin model optimization
+        from the best of a random sample from these parameters' priors. For trainable parameters
+        without priors, we begin optimization from their initial values.
+
+        :param dataset: The data with which to optimize the `model`.
+        """
+
+        num_trainable_params_with_priors = tf.reduce_sum(
+            [
+                tf.size(param)
+                for param in self.model.kernel.trainable_parameters
+                if param.prior is not None
+            ]
+        )
+
+        if num_trainable_params_with_priors >= 1:  # Find a promising kernel initialization
+            num_random_samples = tf.minimum(1000, 100 * num_trainable_params_with_priors)
+            self.find_best_kernel_initialization(num_random_samples)
+
+        self.optimizer.optimize(self.model, dataset)
+
+    def find_best_kernel_initialization(self, num_kernel_samples: int = 100) -> None:
+        """
+        Test `num_kernel_samples` models with kernel parameters sampled from their
+        priors. The model's kernel parameters are then set to those achieving maximal
+        likelihood across the sample.
+
+        :param num_kernel_samples: Number of randomly sampled kernels to evaluate.
+        """
+
+        @tf.function
+        def evaluate_likelihood_of_kernel_parameters() -> tf.Tensor:
+            return self.model.maximum_log_likelihood_objective()
+
+        current_best_parameters = read_values(self.model.kernel)
+        max_log_likelihood = evaluate_likelihood_of_kernel_parameters()
+
+        for _ in tf.range(num_kernel_samples):
+            randomize_kernel_hyperparameters(self.model.kernel)
+            try:
+                likelihood = evaluate_likelihood_of_kernel_parameters()
+            except tf.errors.InvalidArgumentError:  # allow badly specified priors
+                likelihood = -1e100
```

Comment:
Why not 0?

Reply:
Ah because this is really the log_likelihood

Reply:
made this clear